Add endpoints for getting/setting workspace permissions

jjnesbitt opened the pull request:
Fixes multinet-app/multinet-server#419 
Depends on multinet-app/multinet-server#436 

This adds endpoints to retrieve and set the permissions on a workspace. It, however, does not yet allow for the changing of the owner of a workspace, as that is intended to be a separate action. It also adds a temporary transformation of the underlying permission data (f20bc1f), which is needed for the client to function properly. This temporary fix is intended to be rolled back once the [ORM](https://github.com/multinet-app/multinet-rfcs/pull/12) is implemented, as that's a natural point to change how we're storing this data.

jjnesbitt committed:
Use class syntax for TypedDict

jjnesbitt committed:
Use class syntax for TypedDict

jjnesbitt committed:
Add set_workspace_permissions db function

jjnesbitt committed:
Add set_workspace_permissions db function

jjnesbitt committed:
Add/update workspace permission endpoints

jjnesbitt committed:
Add/update workspace permission endpoints

jjnesbitt committed:
Add comments

jjnesbitt committed:
Add comments

AlmightyYakob commented on .github/workflows/main.yml:
```diff
@@ -32,11 +32,15 @@ jobs:
         run: diff requirements.txt generated-requirements.txt
 
+      - name: Capture pipenv version
+        run: echo "::set-output name=pipenv-version::$(pipenv --version | sed 's/,//g')"
```
First off, thanks for doing this.

I think if we're gonna bother using `sed`, why not just remove all of the output that isn't the version anyway.

```suggestion
        run: echo "::set-output name=pipenv-version::$(pipenv --version | sed 's/pipenv, version //g')"
```

Also, why is the `::set-output` needed here?

AlmightyYakob replied:
Also, I think once everything here is resolved, we can probably squash these two commits into one.

waxlamp replied:
The reason I just removed the comma was that, if there's any new release of Pipenv at all and the version string changes, I want the cache to be invalidated. I would have left the comma there except that GH Actions doesn't allow them in cache hash strings. It is probably low-risk to do it your way, but I don't see any real advantages to it. As I said, my preference was to just leave the version string raw, but the comma gums that up. What do you think?

That `::set-output...::` thing is a GH Actions "workflow command". You can inject certain commands through an `echo` invocation to make certain things happen. `set-output` causes whatever comes after the "closing" `::` to appear in the `steps.<id>.outputs.<output-name>` global property, and it becomes available in subsequent steps. Here I used that to inject the version string into the cache hash value. Here's the documentation for [workflow commands](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions). Very cool stuff!

waxlamp replied:
> Also, I think once everything here is resolved, we can probably squash these two commits into one.

Agreed. I meant to mention this in a comment but forgot. When you're happy with everything, feel free to do an interactive rebase to squash those commits as needed (or, if you're feeling very pedantic, you could cherry pick the combined commit to another PR, then rebase this on top of that--but I don't really think that's necessary).

AlmightyYakob replied:
> The reason I just removed the comma was that, if there's any new release of Pipenv at all and the version string changes, I want the cache to be invalidated. I would have left the comma there except that GH Actions doesn't allow them in cache hash strings. It is probably low-risk to do it your way, but I don't see any real advantages to it. As I said, my preference was to just leave the version string raw, but the comma gums that up. What do you think?

The main reason I suggested it is because currently there's still a space in the cache string. It's not a huge deal, and I guess is allowed, but that was the main reason.

> That `::set-output...::` thing is a GH Actions "workflow command". You can inject certain commands through an `echo` invocation to make certain things happen. `set-output` causes whatever comes after the "closing" `::` to appear in the `steps.<id>.outputs.<output-name>` global property, and it becomes available in subsequent steps. Here I used that to inject the version string into the cache hash value. Here's the documentation for [workflow commands](https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions). Very cool stuff!

:+1:

waxlamp replied:
> > The reason I just removed the comma was that, if there's any new release of Pipenv at all and the version string changes, I want the cache to be invalidated. I would have left the comma there except that GH Actions doesn't allow them in cache hash strings. It is probably low-risk to do it your way, but I don't see any real advantages to it. As I said, my preference was to just leave the version string raw, but the comma gums that up. What do you think?
> 
> The main reason I suggested it is because currently there's still a space in the cache string. It's not a huge deal, and I guess is allowed, but that was the main reason.
> 

Since spaces are allowed, I am not worried about them remaining in the string. I'm more worried about a future version  update *changing* the format of the version output (and thus making our postprocessing logic broken somehow), so I made the minimal change to the string to get it to work (i.e., zapping the commas).

jjnesbitt committed:
Add function that uses sub value for user lookup

jjnesbitt committed:
Merge branch 'workspace-permission-endpoints' of github.com:multinet-app/multinet-server into workspace-permission-endpoints

jjnesbitt committed:
Add workspace permisssion user object transform

These functions are temporary solutions to the problem of our permission
storage. Right now, we only store the `sub` value of a user, which means
not all of the information we need is available on the client. The
future solution to this is to store whole user objects in the
`permissions` field of a workspace. However, doing that now would
conflict with future work, so this is the temporary solution.

jjnesbitt committed:
Merge branch 'master' into workspace-permission-endpoints

jjnesbitt committed:
Remove cache in `get_workspace_metadata`

waxlamp commented on multinet/api.py:
```diff
@@ -25,11 +27,62 @@
 
 
+# Included here due to circular imports
+# TODO: Remove once implementing new ORM and permission storage
```
Is your plan to finish the ORM work first, then rebase on (or merge from) that work here?

AlmightyYakob replied:
I figured that this would make it in first, so the comment was there to leave a breadcrumb for circling back. Do you think it's worth waiting for the ORM work to make it in first? Managing permissions from the client relies on this endpoint.

waxlamp replied:
If the circular dependency is causing test failures, I'm not sure that we should simply merge it in. If there's an easy way to fix that issue, then we should do it. If that will cause a major headache then we may need to go out on a limb and merge in "broken" code but I'd like to avoid that if possible.

What do you think?

AlmightyYakob replied:
The main reason I didn't solve the circular dependency issue in a more robust way is because it seemed a bit out of scope for this issue, and the functions that are causing them to happen (these conversion functions) are temporary and would be removed when the ORM change comes down. Also, if we encounter this same issue during the ORM refactor, that's a perfectly valid time to fix it. I just didn't want to spend a ton of time reorganizing the code base just for it not to matter once we ultimately reorganize it for the ORM.

waxlamp replied:
I agree with what you're saying. But this will not work if the server can't actually run because of the circular import. If you can certify that this does work properly, then we're all good. I would just ask you to file an issue pointing to these code blobs (a TODO in the codebase isn't quite as trackable as I'd like).

AlmightyYakob replied:
> I agree with what you're saying. But this will not work if the server can't actually run because of the circular import. If you can certify that this does work properly, then we're all good.

This does work properly (as in, no circular imports at runtime), although having someone else test this out may be a good idea.

> I would just ask you to file an issue pointing to these code blobs (a TODO in the codebase isn't quite as trackable as I'd like).

#442

jjnesbitt committed:
Add small json validation and comment

jjnesbitt committed:
Merge branch 'master' into workspace-permission-endpoints

jjnesbitt committed:
Add cast to `WorkspacePermissions`

jjnesbitt committed:
Merge branch 'workspace-permission-endpoints' of github.com:multinet-app/multinet-server into workspace-permission-endpoints

jjnesbitt committed:
Update tests

waxlamp commented on multinet/api.py:
```diff
@@ -24,11 +27,62 @@
+# Included here due to circular imports
+# TODO: Remove once implementing new ORM and permission storage
+def _permissions_id_to_user(permissons: WorkspacePermissions) -> Dict:
+    """Transform permission documents to directly container user info."""
```
The grammar here is a bit off. Should it be `contain`?

waxlamp replied:
I think a longer docstring would be useful here. Reading the source code is somewhat confusing due to the types shifting and the structure of our basic data changing. A paragraph following the summary comment you have here would be great because newcomers to the codebase (in this case, me!) would better understand what this utility function is for, and why it's written in this way, etc.

AlmightyYakob replied:
c60d404

waxlamp commented on multinet/api.py:
```diff
@@ -24,11 +27,62 @@
+            continue
+
+        if role == "owner":
+            assert isinstance(users, str)
```
We shouldn't have any `assert` statements in the main code. What role is it serving here?

waxlamp replied:
Ah I think I might understand. Is this something `mypy` makes use of to narrow from Union types?

AlmightyYakob replied:
Yeah, it's to deal with `mypy`. Since all other fields besides `public` are arrays, we have to tell `mypy` that in this case, it's just a string, since `mypy` is too stupid to figure that out on its own.

waxlamp commented on multinet/api.py:
```diff
@@ -24,11 +27,62 @@
+
+            user = find_user_from_id(users)
+            if user is not None:
+                # Not sure why mypy complains about `user` here
```
What is the exact type error you get?

AlmightyYakob replied:
It was giving me an error regarding user being a `str` or something. However I'm actually about to push a change which will fix the type errors, so this will be moot.

waxlamp commented on multinet/api.py:
```diff
@@ -24,11 +27,62 @@
+            user = find_user_from_id(users)
+            if user is not None:
+                # Not sure why mypy complains about `user` here
+                new_permissions["owner"] = asdict(user)  # type: ignore
```
What's this actually doing? It's placing the full user object in place of the sub string?

AlmightyYakob replied:
Exactly

waxlamp commented on multinet/api.py:
```diff
@@ -24,11 +27,62 @@
+
+            new_users = []
+            for sub in users:
+                user = find_user_from_id(sub)
```
Does this make the loop O(N^2)? Is that ok?

AlmightyYakob replied:
I don't think it's a problem, since this it's really O(n*m), where n is the number of keys in the permissions dict (5), and m is the number of users in each role (could just use the role with the most users, to be safe). So it's really linear.

waxlamp commented on multinet/api.py:
```diff
@@ -24,11 +27,62 @@
+            # Ignoring requirement of literal string keys
+            new_permissions[role] = new_users  # type: ignore
+
+    return cast(Dict, new_permissions)
```
Is `WorkspacePermissions` a dataclass? How does this `cast` operator work to turn that into a `Dict`?

AlmightyYakob replied:
It's a `TypedDict`, and I'm not sure how the `cast` operator works internally, but it's used here to remove the restrictions of it being a `WorkspacePermissions` dict, because it isn't anymore, due to the replacements above. I'm actually about to push a change where I move this operation, but it'll still be in this function.

waxlamp commented on multinet/api.py:
```diff
@@ -41,14 +95,32 @@ def get_workspaces() -> Any:
     metadata = db.get_workspace_metadata(workspace)
 
-    return {"name": metadata["name"], "permissions": metadata["permissions"]}
+    return _permissions_id_to_user(metadata["permissions"])
```
This seems to be the only call to `_permissions_id_to_user`. Why not inline the function here?

AlmightyYakob replied:
I just preferred to keep those temporary functions clearly separated and marked from the rest of the api. Since eventually, we'll remove those functions, but the api endpoint will remain.

waxlamp commented on multinet/api.py:
```diff
@@ -41,14 +95,32 @@ def get_workspaces() -> Any:
+    }:
+        raise MalformedRequestBody(request.json)
+
+    perms = _permissions_user_to_id(request.json)
```
Same question as above about inlining a one-time-use function.

AlmightyYakob replied:
Same response as above

waxlamp commented on multinet/api.py:
```diff
@@ -41,14 +95,32 @@ def get_workspaces() -> Any:
+        "readers",
+        "public",
+    }:
+        raise MalformedRequestBody(request.json)
```
I think I might disagree with this. A `PUT` call usually means that you specify only what you want changed, and the server figures out how to apply your "diff".

This way of operating also works, but there's a danger: suppose you `GET` the permissions, then you have faulty logic for updating that object so that you accidentally put in changes you didn't want. Then on `PUT`, you will end up with bad state.

This is probably a weak argument. I'm willing to hear why it should be required to pass back the whole permissions object. One argument that feels at least somewhat compelling is that the server logic could become more complex to handle partial updates.

In any case, there's no need to change things in this PR. I'd like to have the discussion and if we decide that partial `PUT`s are warranted, we'll just file an issue and worry about it later.

AlmightyYakob replied:
> I think I might disagree with this. A `PUT` call usually means that you specify only what you want changed, and the server figures out how to apply your "diff".

I disagree, I think of `PUT` as an updating an existing resource to match what you specify (more or less). You seem to be describing the `PATCH` method.


> This way of operating also works, but there's a danger: suppose you `GET` the permissions, then you have faulty logic for updating that object so that you accidentally put in changes you didn't want. Then on `PUT`, you will end up with bad state.

I think this could happen in any way of utilizing the endpoint. At the end of the day, it's our responsibility to make sure the client uses this endpoint properly. Plus, you still can't modify owners from these endpoints, so in the worst case the owner can still fix things after an error.

> This is probably a weak argument. I'm willing to hear why it should be required to pass back the whole permissions object. One argument that feels at least somewhat compelling is that the server logic could become more complex to handle partial updates.

Yeah this is one of the main reasons, and I don't really see a benefit to specifying only the change rather than the whole new object. It makes the logic on the server more complex, and it also makes the logic on the client much more complex.

waxlamp replied:
You are right, especially in pointing out that I'm thinking of `PATCH` rather than `PUT`. I agree with everything you said here.

waxlamp commented on multinet/db.py:
```diff
@@ -222,13 +229,38 @@ def get_workspace_metadata(name: str) -> Workspace:
     # Find the metadata record for the named workspace. If it's not there,
     # something went very wrong, so bail out.
-    metadata = workspace_mapping(name)
+    metadata = workspace_mapping.__wrapped__(name)
```
Remind me what `.__wrapped__()` does here? It just calls the undecorated version of the function?

Why do we need that here? If you include a comment explaining the need, that would be good.

(Hopefully this becomes moot after the ORM stuff lands but still, in the meantime we should cover our donkeys.)

AlmightyYakob replied:
Yeah it calls the un-decorated function. The reason I added that is because when calling `get_workspace_metadata`, we want the most recent version of the metadata, and this function itself isn't cached, so we couldn't just call the un-decorated version of `get_workspace_metadata`, as it doesn't exist. With the changes that I said I'd be making in my other comment, this might actually not be necessary anymore.

I'd still be okay with using the un-cached function, as long as it doesn't impact performance. What do you think?

waxlamp replied:
Correctness matters more than performance in this case, so do whatever makes the most sense. I'd just include a comment about why it's this way, just to spare confusion or errors in the future.

AlmightyYakob replied:
2b31899 I undid the use of the wrapped function, so I didn't think a comment was necessary anymore

waxlamp commented on multinet/db.py:
```diff
@@ -222,13 +229,38 @@ def get_workspace_metadata(name: str) -> Workspace:
+
+    doc = copy.deepcopy(workspace_mapping(name))
+    if doc is None:
+        raise DatabaseCorrupted()
```
We should think about including more information in the `DatabaseCorrupted` exception. Even just an error message.

I'll file an issue for this.

waxlamp replied:
See #443.

waxlamp commented on multinet/db.py:
```diff
@@ -222,13 +229,38 @@ def get_workspace_metadata(name: str) -> Workspace:
+    if doc is None:
+        raise DatabaseCorrupted()
+
+    # TODO: Do user object validation once ORM is implemented
```
Can you make sure to file an issue for this TODO?

AlmightyYakob replied:
https://github.com/multinet-app/multinet-server/issues/444

waxlamp commented on multinet/db.py:
```diff
@@ -222,13 +229,38 @@ def get_workspace_metadata(name: str) -> Workspace:
+
+    # TODO: Do user object validation once ORM is implemented
+
+    # To ensure owner is only changed explicitly
```
What do you mean by this?

AlmightyYakob replied:
I mean that the owner can't be changed through setting the workspace permissions like everything else, it needs to be explicitly changed through it's own endpoint or something.

waxlamp replied:
Oh I see. I'd suggest different wording for the comment:

```suggestion
    # Disallow changing workspace ownership through this function.
```

waxlamp commented on multinet/db.py:
```diff
@@ -222,13 +229,38 @@ def get_workspace_metadata(name: str) -> Workspace:
+
+    doc["permissions"] = new_permissions
+    return_doc = workspace_mapping_collection().get(
+        workspace_mapping_collection().update(doc, check_rev=False)
```
Why `check_rev=False`?

waxlamp replied:
Also, what does the `update()` method return? I'm surprised it doesn't give you back the new doc!

AlmightyYakob replied:
Yeah, `update` only returns the metadata (so `_rev`, `_id`, etc.). The reason I added `check_rev=False`, is because I was getting a `DocumentRevisionError`, which was due to the call to `workspace_mapping` above returning an old version of the document, with an old `_rev` (since it's cached). 

Looking at this now, I actually don't like what I did, as it just ignores the problem. I'm going to push a fix that actually addresses the problem here.

waxlamp replied:
> Looking at this now, I actually don't like what I did, as it just ignores the problem. I'm going to push a fix that actually addresses the problem here.

That's why I spoke up, something felt off about setting that flag.

AlmightyYakob replied:
6a786b4

waxlamp commented on multinet/swagger/get_workspace_permissions.yaml:
```diff
@@ -1,4 +1,4 @@
-Retrieve a single workspace.
+Retrieve the permissions on a workspace.
```
Doesn't more in this file need to be updated?

AlmightyYakob replied:
98a0c81

waxlamp commented on multinet/swagger/set_workspace_permissions.yaml:
```diff
@@ -0,0 +1,22 @@
+  200:
+    description: Metadata about the specified workspace
+    schema:
+      $ref: "#/definitions/workspace"
```
If `#/definitions/workspace` now details workspace permissions, its name should probably change to reflect that.

AlmightyYakob replied:
98a0c81

waxlamp commented on multinet/types.py:
```diff
@@ -5,16 +5,15 @@
-    },
-)
+
+class WorkspacePermissions(TypedDict):
```
Why still inherit from `TypedDict`? I thought you were on a mission to exterminate them?

AlmightyYakob replied:
I am still on a mission, I just changed the declaration of `WorkspacePermissions` to use the class based syntax, as it's more clear.

waxlamp commented on multinet/types.py:
```diff
@@ -5,16 +5,15 @@
+class WorkspacePermissions(TypedDict):
+    """Permissions on a Workspace."""
+
+    owner: str  # this is the `sub` property of the `User` dataclass
```
I wonder if we should just declare a type alias, `UserSub` that equals `str`. Just something to think about. Maybe it's more trouble than it's worth.

AlmightyYakob replied:
Perhaps, but I think the with the new ORM changes it would become moot.

waxlamp commented on test/test_meta.py:
```diff
@@ -7,7 +7,7 @@ def test_generated_workspace(managed_workspace, managed_user, server):
 
     with conftest.login(managed_user, server):
-        resp = server.get(f"/api/workspaces/{managed_workspace}")
+        resp = server.get(f"/api/workspaces/{managed_workspace}/permissions")
```
Shouldn't we check that the permissions are what we expect?

Perhaps we need a bit more complex fixtures to make this happen easily. I have some ideas.

It's ok to not address this now, but please file an issue instead.

waxlamp replied:
Oh I see, you ran that test below. n/m.

AlmightyYakob replied:
Yeah, this just updates the endpoint with the new URL.

waxlamp reviewed (changes requested):
This looks great, thanks!

I left a whole bunch of questions and a couple of minor suggestions.

jjnesbitt committed:
Move cast operation, fix type errors, add comments

jjnesbitt committed:
Fix DocumentRevisionError when setting permissions

jjnesbitt committed:
Update comment regarding ownership

Co-authored-by: Roni Choudhury <2903332+waxlamp@users.noreply.github.com>

jjnesbitt committed:
Update docstring for `_permissions_id_to_user`

jjnesbitt committed:
Merge branch 'workspace-permission-endpoints' of github.com:multinet-app/multinet-server into workspace-permission-endpoints

jjnesbitt committed:
Update permissions swagger docs

jjnesbitt committed:
Remove lingering assert, fix return of set perms

jjnesbitt committed:
Remove use of wrapped function
